extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -237,6 +261,101 @@ export class AmpAdExit extends AMP.BaseElement {
       this.user().error(TAG, 'Invalid JSON config', e);
       throw e;
     }
+
+    this.ampAdResourceId_ = user().assert(
+        this.getAmpAdResourceId_(),
+        `${TAG}: No friendly parent amp-ad element was found for amp-ad-exit.`);
+
+    this.init3pResponseListener_();
+  }
+
+  /**
+   * Gets the resource ID of the amp-ad element containing this amp-ad-exit tag.
+   * This is a pass-through for the version in service.js, solely because
+   * the one in service.js isn't stubbable for testing, since only object
+   * methods are stubbable.
+   * @returns {?string}
+   * @private
+   * @VisibleForTesting
+   */
+  getAmpAdResourceId_() {
+    return getAmpAdResourceId(this.element, this.win.top);
+  }
+
+  /** @override */
+  resumeCallback() {
+    this.init3pResponseListener_();
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.unlisten_) {
+      this.unlisten_();
+      this.unlisten_ = null;
+    }
+    return super.unlayoutCallback();
+  }
+
+  /**
+   * amp-analytics will create an iframe for vendors in
+   * extensions/amp-analytics/0.1/vendors.js who have transport/iframe defined.
+   * This is limited to MRC-accreddited vendors. The frame is removed in
+   * amp-analytics, and the listener is destroyed here, if the user
+   * navigates/swipes away from the page. Both are recreated if the user
+   * navigates back to the page.
+   * @private
+   */
+  init3pResponseListener_() {
+    this.unlisten_ = this.unlisten_ || listen(this.getAmpDoc().win, 'message',
```

Comment:
It seems like you're trying to guard for either

- Resume before Build (which should never happen)
- Unlayout which doesn't actually unlayout (so layout is never called again)?

Reply:
This was originally in detachedCallback(), but previous review comments / in-person meeting said to put it into unlayoutCallback(), for the use case in which the user swipes one (and only one) story away in the carousel. @keithwrightbos can provide more insight.

Reply:
I haven't checked this in yet, but how about I remove "this.unlisten_ ||" from the line above, but insert the following line before:
```
    dev().assert(!this.unlisten_, 'Unlistener should not already exist.');
```

Reply:
(Checked that in)